SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_MaterialFarmer.cpp

```diff
@@ -0,0 +1,566 @@
+          LockMode::UNLOCK_WHILE_RUNNING,
+          3
+    )
+    , LANGUAGE(
+        "<b>Game Language:</b><br>Required to read sandwich ingredients.",
+        IV_READER().languages(),
+        LockMode::UNLOCK_WHILE_RUNNING,
+        false
+    )
+    , SANDWICH_OPTIONS(LANGUAGE)
+    , GO_HOME_WHEN_DONE(true)
+    , AUTO_HEAL_PERCENT(
+        "<b>Auto-Heal %</b><br>Auto-heal if your HP drops below this percentage.",
+        LockMode::UNLOCK_WHILE_RUNNING,
+        75, 0, 100
+    )
+    , SAVE_DEBUG_VIDEO(
+        "<b>DEV MODE: Save debug videos to Switch:</b><br>"
+        "Set this on to save a Switch video everytime an error occurs. You can send the video to developers to help them debug later.",
+        LockMode::LOCK_WHILE_RUNNING,
+        false
+    )
+    , SKIP_WARP_TO_POKECENTER(
+        "<b>DEV MODE: Skip warping to closest PokeCenter:</b><br>"
+        "This is for debugging the program without waiting for the initial warp.",
+        LockMode::LOCK_WHILE_RUNNING,
+        false
+    )
+    , SKIP_SANDWICH(
+        "<b>DEV MODE: Skip making sandwich:</b><br>"
+        "This is for debugging the program without waiting for sandwich making.",
+        LockMode::UNLOCK_WHILE_RUNNING,
+        false
+    )
+    , NOTIFICATION_STATUS_UPDATE("Status Update", true, false, std::chrono::seconds(3600))
+    , NOTIFICATIONS({
+        &NOTIFICATION_STATUS_UPDATE,
+        &NOTIFICATION_PROGRAM_FINISH,
+        &NOTIFICATION_ERROR_RECOVERABLE,
+        &NOTIFICATION_ERROR_FATAL,
+    })
+{
+    if (PreloadSettings::instance().DEVELOPER_MODE){
+        PA_ADD_OPTION(SAVE_DEBUG_VIDEO);
+        PA_ADD_OPTION(SKIP_WARP_TO_POKECENTER);
+        PA_ADD_OPTION(SKIP_SANDWICH);
+    }
+    PA_ADD_OPTION(SAVE_GAME_BEFORE_SANDWICH);
+    PA_ADD_OPTION(NUM_SANDWICH_ROUNDS);
+    PA_ADD_OPTION(LANGUAGE);
+    PA_ADD_OPTION(SANDWICH_OPTIONS);
+    PA_ADD_OPTION(GO_HOME_WHEN_DONE);
+    PA_ADD_OPTION(AUTO_HEAL_PERCENT);
+    PA_ADD_OPTION(NOTIFICATIONS);
+}
+
+
+void MaterialFarmer::program(SingleSwitchProgramEnvironment& env, BotBaseContext& context){
+    MaterialFarmer_Descriptor::Stats& stats = env.current_stats<MaterialFarmer_Descriptor::Stats>();
+
+
+    assert_16_9_720p_min(env.logger(), env.console);
+
+    //  Connect the controller.
+    pbf_press_button(context, BUTTON_L, 10, 50);
+
+    // Throw user setup errors early in program
+    // - Ensure language is set
+    const Language language = LANGUAGE;
+    if (language == Language::None) {
+        throw UserSetupError(env.console.logger(), "Must set game language option to read ingredient lists.");
+    }
+
+    // - Ensure audio input is enabled
+    LetsGoKillSoundDetector audio_detector(env.console, [](float){ return true; });
+    wait_until(
+        env.console, context,
+        std::chrono::milliseconds(1100),
+        {
+            static_cast<AudioInferenceCallback&>(audio_detector)
+        }
+    );
+    audio_detector.throw_if_no_sound(std::chrono::milliseconds(1000));
```

Comment:
Does this need to happen explicitly at the beginning here? Or can be done as part of whatever uses sound later?

Reply:
It can be done later. But I thought it would be better from a user experience point of view if the user setup error was thrown early, as opposed to 5-10 minutes later, when the user may have stepped away.

Do you feel it's better to do this check later?